internal/pkg/graph/graph.go

```diff
@@ -17,6 +24,8 @@ const (
 type Graph[V comparable] struct {
 	vertices  map[V]neighbors[V] // Adjacency list for each vertex.
 	inDegrees map[V]int          // Number of incoming edges for each vertex.
+	status    map[V]string       // status of each vertex.
+	lock      sync.Mutex         // lock used to mutate graph data.
```

Comment:
I modified the code where every method of `LabeledGraph[V comparable]` struct uses `lock `.

Couple of reasons why i want to use lock declaring over here instead of caller.
1. If the lock is managed within the data structure, it simplifies the usage for the caller. The caller does not need to worry about acquiring and releasing the lock, making the API easier to use
2. The caller may not have full context or understanding of the graph's internals, which increases the risk of incorrect or insufficient locking, leading to potential data races or deadlocks.